pin/gc/gc.go

```diff
@@ -35,144 +54,159 @@ type Result struct {
 //
 func GC(ctx context.Context, bs bstore.GCBlockstore, ls dag.LinkService, pn pin.Pinner, bestEffortRoots []*cid.Cid) <-chan Result {
 
-	elock := log.EventBegin(ctx, "GC.lockWait")
-	unlocker := bs.GCLock()
-	elock.Done()
-	elock = log.EventBegin(ctx, "GC.locked")
-	emark := log.EventBegin(ctx, "GC.mark")
-
 	ls = ls.GetOfflineLinkService()
-
 	output := make(chan Result, 128)
+	tri := newTriset()
+
+	emark := log.EventBegin(ctx, "GC.mark")
+
+	err := pn.Flush()
+	if err != nil {
+		output <- Result{Error: err}
+		close(output)
+		return output
+	}
+	addRoots(tri, pn, bestEffortRoots)
 
 	go func() {
```

Comment:
this goroutine should probably be its own function now

Reply:
I am working on refactoring the GC to not be just a function but to have structure, so it will be changed.